lib/Dialect/Triton/Transforms/Combine.cpp

```diff
@@ -49,6 +49,52 @@ DenseElementsAttr getConstantValue(Builder &builder, Attribute value,
   return res;
 }
 
+bool isAddPtrOffsetCombinable(Value first, Value second) {
+  auto GetConstantIntValue = [](Value val) -> std::optional<llvm::APInt> {
+    DenseElementsAttr constAttr;
+    Attribute attr;
+    auto defOp = val.getDefiningOp();
+    if (auto splatOp = llvm::dyn_cast_or_null<SplatOp>(defOp))
+      val = splatOp.getSrc();
+    else if (matchPattern(defOp, m_Constant(&constAttr)) && constAttr.isSplat())
+      attr = constAttr.getSplatValue<Attribute>();
+
+    // Check constant value.
+    llvm::APInt intVal;
+    if (matchPattern(val, m_ConstantInt(&intVal)))
+      return intVal;
+
+    // Check IntegerAttr
+    if (auto intAttr = dyn_cast_or_null<IntegerAttr>(attr))
+      return intAttr.getValue();
+
+    return std::nullopt;
+  };
+
+  Type elem1Type = first.getType();
+  Type elem2Type = second.getType();
+  if (elem1Type.isa<ShapedType>() && elem2Type.isa<ShapedType>()) {
+    elem1Type = elem1Type.cast<ShapedType>().getElementType();
+    elem2Type = elem2Type.cast<ShapedType>().getElementType();
+  }
+
+  if (elem1Type == elem2Type) {
+    // Whether bitwidth of element type is equal to pointer
+    if (elem1Type.getIntOrFloatBitWidth() == 64)
+      return true;
+
+    // first + second does not overflow
+    auto firstVal = GetConstantIntValue(first);
+    auto secondVal = GetConstantIntValue(second);
+    if (firstVal && secondVal) {
+      bool Overflow = false;
+      auto resVal = firstVal->uadd_ov(*secondVal, Overflow);
```

Comment:
I believe we care about signed overflow and not unsigned as offset as sign extended